annotation-processing/src/main/java/com/exactpro/epfast/annotation/processing/FastProcessor.java

```diff
@@ -19,112 +23,198 @@
+        for (Map.Entry<String, Element> entry : fastTypesByName.entrySet()) {
+            PackageElement enclosingPackage = elementUtils.getPackageOf(entry.getValue());
+            while (!packageElementsMap.containsKey(enclosingPackage)) {
+                enclosingPackage = getParent(enclosingPackage);
+            }
+            packageElementsMap.get(enclosingPackage).put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private PackageElement getParent(Element enclosing) {
+        String currentPackage = enclosing.toString();
+        int lastDotIndex = currentPackage.lastIndexOf('.');
+        if (lastDotIndex  == -1) {
+            return elementUtils.getPackageElement("");
+        }
+        return elementUtils.getPackageElement(currentPackage.substring(0, lastDotIndex));
+    }
+
+    private HashMap<PackageElement, HashMap<String, Element>> collectFastPackages(RoundEnvironment roundEnv) {
+        HashMap<PackageElement, HashMap<String, Element>> packageElementMap = new HashMap<>();
+        PackageElement topPackage = elementUtils.getPackageElement("");
+        packageElementMap.put(topPackage, new HashMap<>());
+        Set<? extends Element> packageSet = roundEnv.getElementsAnnotatedWith(fastPackageElement);
+        for (Element element : packageSet) {
+            String packageName = element.getAnnotation(FastPackage.class).name();
+            if (packageName.isEmpty()) {
+                PackageElement packageElement = elementUtils.getPackageOf(element);
+                packageElementMap.put(packageElement, new HashMap<>());
+            } else {
+                try {
+                    PackageElement packageElement = elementUtils.getPackageElement(packageName);
+                    packageElementMap.put(packageElement, new HashMap<>());
+                } catch (Exception e) {
+                    messager.printMessage(
+                        Diagnostic.Kind.ERROR,
+                        String.format("Package for FastPackage name = %s doesn't exist", packageName),
+                        element);
+                }
+            }
+        }
+        return packageElementMap;
+    }
+
     private HashMap<String, Element> collectFastTypes(RoundEnvironment roundEnv) {
-        HashMap<String, ArrayList<Element>> nameClassMap = new HashMap<>();
+        HashMap<String, ArrayList<Element>> fastTypesByName = new HashMap<>();
         Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(FastType.class);
         for (Element element : annotatedElements) {
             String fastTypeName = element.getAnnotation(FastType.class).name();
             if (fastTypeName.isEmpty()) {
                 fastTypeName = element.getSimpleName().toString();
             }
-            nameClassMap.computeIfAbsent(fastTypeName, key -> new ArrayList<>()).add(element);
+            fastTypesByName.computeIfAbsent(fastTypeName, key -> new ArrayList<>()).add(element);
```

Comment:
It's possible to have types with same Fast name if they belong to different Fast packages.
Add test to validate that